File: ios/Sources/AutoPlugin/AutoPlugin.swift
import Foundation
import Capacitor

@objc(AutoPlugin)
public class AutoPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AutoPlugin"
    public let jsName = "Auto"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isAvailable", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setRootTemplate", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setTransientState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getTransientState", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "sendMessage", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = Auto()

    override public func load() {
        AutoBridge.shared.attach(self)
    }

    deinit {
        AutoBridge.shared.detach(self)
    }

    @objc func isAvailable(_ call: CAPPluginCall) {
        call.resolve([
            "available": true,
            "connected": AutoBridge.shared.connected,
            "platform": "ios"
        ])
    }

    @objc func setRootTemplate(_ call: CAPPluginCall) {
        guard let title = call.getString("title"), !title.isEmpty else {
            call.reject("title is required")
            return
        }

        var rawTemplate: [String: Any] = [
            "title": title,
            "sections": call.getArray("sections") ?? []
        ]
        if let emptyText = call.getString("emptyText") {
            rawTemplate["emptyText"] = emptyText
        }

        guard let template = AutoTemplate.fromDictionary(rawTemplate) else {
            call.reject("title is required")
            return
        }

        AutoBridge.shared.setTemplate(template)
        call.resolve()
    }

    @objc func setState(_ call: CAPPluginCall) {
        guard let key = getStateKey(call) else { return }

        guard let value = call.getObject("value") else {
            call.reject("value is required for key=\(key)")
            return
        }

        AutoBridge.shared.setState(key: key, value: value)
        call.resolve()
    }

    @objc func getState(_ call: CAPPluginCall) {
        guard let key = getStateKey(call) else { return }

        call.resolve(stateResult(key: key, value: AutoBridge.shared.getState(key: key)))
    }

    @objc func removeState(_ call: CAPPluginCall) {
        guard let key = getStateKey(call) else { return }

        AutoBridge.shared.removeState(key: key)
        call.resolve()
    }

    @objc func setTransientState(_ call: CAPPluginCall) {
        guard let key = getStateKey(call) else { return }

        guard let value = call.getObject("value") else {
            call.reject("value is required for key=\(key)")
            return
        }

        AutoBridge.shared.setTransientState(key: key, value: value)
        call.resolve()
    }

    @objc func getTransientState(_ call: CAPPluginCall) {
        guard let key = getStateKey(call) else { return }

        call.resolve(stateResult(key: key, value: AutoBridge.shared.getTransientState(key: key)))
    }

    @objc func sendMessage(_ call: CAPPluginCall) {
        guard let type = call.getString("type"), !type.isEmpty else {
            call.reject("type is required")
            return
        }

        AutoBridge.shared.receiveMessage(type: type, payload: call.getObject("payload"))
        call.resolve()
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }

    func emitEvent(_ eventName: String, data: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.notifyListeners(eventName, data: data, retainUntilConsumed: true)
        }
    }

    private func getStateKey(_ call: CAPPluginCall) -> String? {
        guard let key = call.getString("key"), !key.isEmpty else {
            call.reject("key is required")
            return nil
        }

        guard key != AutoBridge.rootTemplateStateKey else {
            call.reject("key is reserved: \(AutoBridge.rootTemplateStateKey)")
            return nil
        }

        return key
    }

    private func stateResult(key: String, value: [String: Any]?) -> [String: Any] {
        var result: [String: Any] = ["key": key]
        if let value = value {
            result["value"] = value
        }
        return result
    }
}

File: android/src/main/kotlin/app/capgo/auto/AutoPlugin.kt
package app.capgo.auto

import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONObject

@CapacitorPlugin(name = "Auto")
class AutoPlugin : Plugin() {
    private val implementation = Auto()

    override fun load() {
        AutoBridge.configure(context)
        AutoBridge.attach(this)
    }

    override fun handleOnDestroy() {
        AutoBridge.detach(this)
        super.handleOnDestroy()
    }

    @PluginMethod
    fun isAvailable(call: PluginCall) {
        val ret = JSObject().apply {
            put("available", true)
            put("connected", AutoBridge.connected)
            put("platform", "android")
        }
        call.resolve(ret)
    }

    @PluginMethod
    fun setRootTemplate(call: PluginCall) {
        val template = AutoTemplate.fromJson(call.data)
        if (template == null) {
            call.reject("title is required")
            return
        }

        AutoBridge.setTemplate(template)
        call.resolve()
    }

    @PluginMethod
    fun setState(call: PluginCall) {
        val key = getStateKey(call) ?: return

        val value = call.getObject("value")
        if (value == null) {
            call.reject("value is required for key=$key")
            return
        }

        AutoBridge.setState(key, value)
        call.resolve()
    }

    @PluginMethod
    fun getState(call: PluginCall) {
        val key = getStateKey(call) ?: return

        call.resolve(stateResult(key, AutoBridge.getState(key)))
    }

    @PluginMethod
    fun removeState(call: PluginCall) {
        val key = getStateKey(call) ?: return

        AutoBridge.removeState(key)
        call.resolve()
    }

    @PluginMethod
    fun setTransientState(call: PluginCall) {
        val key = getStateKey(call) ?: return

        val value = call.getObject("value")
        if (value == null) {
            call.reject("value is required for key=$key")
            return
        }

        AutoBridge.setTransientState(key, value)
        call.resolve()
    }

    @PluginMethod
    fun getTransientState(call: PluginCall) {
        val key = getStateKey(call) ?: return

        call.resolve(stateResult(key, AutoBridge.getTransientState(key)))
    }

    @PluginMethod
    fun sendMessage(call: PluginCall) {
        val type = call.getString("type")
        if (type.isNullOrBlank()) {
            call.reject("type is required")
            return
        }

        AutoBridge.receiveMessage(type, call.getObject("payload"))
        call.resolve()
    }

    @PluginMethod
    fun getPluginVersion(call: PluginCall) {
        val ret = JSObject().apply {
            put("version", implementation.getPluginVersion())
        }
        call.resolve(ret)
    }

    internal fun emitEvent(eventName: String, data: JSObject) {
        notifyListeners(eventName, data, true)
    }

    private fun getStateKey(call: PluginCall): String? {
        val key = call.getString("key")
        if (key.isNullOrBlank()) {
            call.reject("key is required")
            return null
        }

        if (key == ROOT_TEMPLATE_STATE_KEY) {
            call.reject("key is reserved: $ROOT_TEMPLATE_STATE_KEY")
            return null
        }

        return key
    }

    private fun stateResult(key: String, value: JSONObject?): JSObject {
        return JSObject().apply {
            put("key", key)
            value?.let { put("value", it) }
        }
    }
}
